File: game-engine/src/main/java/org/luxons/sevenwonders/game/boards/Science.java
package org.luxons.sevenwonders.game.boards;

import java.util.Arrays;
import java.util.EnumMap;
import java.util.Map;

public class Science {

    private Map<ScienceType, Integer> quantities = new EnumMap<>(ScienceType.class);

    private int jokers;

    public void add(ScienceType type, int quantity) {
        quantities.merge(type, quantity, (x, y) -> x + y);
    }

    public void addJoker(int quantity) {
        jokers += quantity;
    }

    public int getJokers() {
        return jokers;
    }

    public void addAll(Science science) {
        science.quantities.forEach(this::add);
        jokers += science.jokers;
    }

    public int getQuantity(ScienceType type) {
        return quantities.getOrDefault(type, 0);
    }

    public int size() {
        return quantities.values().stream().mapToInt(q -> q).sum() + jokers;
    }

    public int computePoints() {
        ScienceType[] types = ScienceType.values();
        Integer[] values = new Integer[types.length];
        for (int i = 0; i < types.length; i++) {
            values[i] = quantities.getOrDefault(types[i], 0);
        }
        return computePoints(values, jokers);
    }

    private static int computePoints(Integer[] values, int jokers) {
        if (jokers == 0) {
            return computePointsNoJoker(values);
        }
        int maxPoints = 0;
        for (int i = 0; i < values.length; i++) {
            values[i]++;
            maxPoints = Math.max(maxPoints, computePoints(values, jokers - 1));
            values[i]--;
        }
        return maxPoints;
    }

    private static int computePointsNoJoker(Integer[] values) {
        int independentSquaresSum = Arrays.stream(values).mapToInt(i -> i * i).sum();
        int nbGroupsOfAll = Arrays.stream(values).mapToInt(i -> i).min().orElse(0);
        return independentSquaresSum + nbGroupsOfAll * 7;
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/boards/Science.kt
package org.luxons.sevenwonders.game.boards

enum class ScienceType {
    COMPASS,
    WHEEL,
    TABLET
}

class Science {

    private val quantities: MutableMap<ScienceType, Int> = mutableMapOf()

    var jokers: Int = 0
        private set

    fun size(): Int = quantities.values.sum() + jokers

    fun add(type: ScienceType, quantity: Int) {
        quantities.merge(type, quantity) { x, y -> x + y }
    }

    fun addJoker(quantity: Int) {
        jokers += quantity
    }

    fun addAll(science: Science) {
        science.quantities.forEach { type, quantity -> this.add(type, quantity) }
        jokers += science.jokers
    }

    fun getQuantity(type: ScienceType): Int = quantities.getOrDefault(type, 0)

    fun computePoints(): Int {
        val values = ScienceType.values().map(::getQuantity).toMutableList()
        return computePoints(values, jokers)
    }

    private fun computePoints(values: MutableList<Int>, jokers: Int): Int {
        if (jokers == 0) {
            return computePointsNoJoker(values)
        }
        var maxPoints = 0
        for (i in values.indices) {
            values[i]++
            maxPoints = Math.max(maxPoints, computePoints(values, jokers - 1))
            values[i]--
        }
        return maxPoints
    }

    private fun computePointsNoJoker(values: List<Int>): Int {
        val independentSquaresSum = values.map { i -> i * i }.sum()
        val nbGroupsOfAll = values.min() ?: 0
        return independentSquaresSum + nbGroupsOfAll * 7
    }
}
